src/BasicInputs/DatetimeInput.tsx

```diff
@@ -1,27 +1,153 @@
+'use client';
+
 import {
-  type ComponentPropsWithoutRef,
-  type ElementRef,
   forwardRef,
+  useEffect,
+  useImperativeHandle,
+  useRef,
+  useState,
 } from 'react';
 
-import type { FormControlChildProps } from '@/Form';
-import { Input } from '@/lib/components/ui/input';
+import { add, format } from 'date-fns';
+import { Calendar as CalendarIcon } from 'lucide-react';
+
 import { cn } from '@/lib/utils';
 
+import { Button } from '../lib/components/ui/button';
+import { Calendar } from '../lib/components/ui/calendar';
+import {
+  Popover,
+  PopoverContent,
+  PopoverTrigger,
+} from '../lib/components/ui/popover';
+import {
+  type Granularity,
+  TimePicker12h,
+  TimePicker24h,
+} from '../lib/components/ui/time-picker';
+
+type DatetimeInputRef = {
+  value?: string;
+} & Omit<HTMLButtonElement, 'value'>;
+
+export interface DatetimeInputProps {
+  value?: string;
+  onChange?: (date: string | undefined) => void;
+  granularity?: Granularity;
+  hourCycle?: 12 | 24;
+}
+
 export const DatetimeInput = forwardRef<
-  ElementRef<typeof Input>,
-  FormControlChildProps & ComponentPropsWithoutRef<typeof Input>
->(({ value = '', className, ...inputProps }, ref) => {
-  if (typeof value !== 'string') {
-    throw new Error('value must be of type string.');
-  }
-  return (
-    <Input
-      type="datetime-local"
-      value={value?.substring(0, 19) ?? ''}
-      className={cn('dark:[color-scheme:dark]', className)}
-      ref={ref}
-      {...inputProps}
-    />
-  );
-});
+  Partial<DatetimeInputRef>,
+  DatetimeInputProps
+>(
+  (
+    { value, onChange, granularity = 'Second', hourCycle = 12, ...props },
+    ref,
+  ) => {
+    const [date, setDate] = useState<Date | undefined>(
+      value && !Number.isNaN(new Date(value).valueOf())
+        ? new Date(value)
+        : undefined,
+    );
+    useEffect(() => {
+      onChange?.(date ? date.toISOString() : undefined);
+    }, [date, onChange]);
```

Comment:
Controlled inputs should never store the value as `undefined`, set it to null instead. 

Thoughts on just making a `dateToISOString` fn that accepts `Date | null` and does the appropriate parsing inside?
Basically replacing every location where you call setDate with this fn. This would eliminate the need to use react hooks. You can also just define the `date` variable normally without useState since the onChange fn would change it and the new value is passed as a prop instead.

Reply:
Removed the unnecessary useState. 

One thing to note about the clear functionality is that calling onChange(undefined) does not mean we will see `Pick a Date` in the component. Instead, we will see the initial date that was passed into the component. However, calling onChange(null) fixes this problem.

Since most of the functions/components I imported only expect type Date|undefined, I added some lines of code that convert between null and undefined in DateTimeInput so that we get the expected interaction and no typing errors. I am not sure if this is the best approach to this problem so if you have any ideas, let me know.